lib/url-parse.js

```diff
@@ -1,58 +1,111 @@
 var url = require('url')
-  , encode = function (value) {
-    return encodeURIComponent(value).replace(/[!'()*]/g, function (c) {
-      return '%' + c.charCodeAt(0).toString(16).toUpperCase()
+  , STRING = 'string'
+  , stringify
+  , parse
+  , encode
+
+/**
+ * Percent encodes a query string according to RFC 3986
+ *
+ * @param value
+ * @returns {string}
+ */
+encode = function (value) {
+  return encodeURIComponent(value).replace(/[!'()*]/g, function (c) {
+    return '%' + c.charCodeAt(0).toString(16).toUpperCase()
+  })
+}
+
+/**
+ * Parses a query string into an array, preserving parameter values
+ *
+ * @param string
+ * @returns {*}
+ */
+parse = function (string) {
+  var parts
+  if (typeof string === STRING) {
+    parts = string.split('&')
+    return parts.map(function (param) {
+      var index = (typeof param === STRING) ? param.indexOf('=') : -1,
+        paramObj = {}
+
+      // this means that there was no value for this key (not even blank, so we store this info) and the value is set
+      // to null
+      if (index < 0) {
+        paramObj.key = param.substr(0, param.length)
+        paramObj.value = null
+      }
+      else {
+        paramObj.key = param.substr(0, index)
+        paramObj.value = param.substr(index + 1)
+      }
+
+      return paramObj
     })
   }
-  , stringify = function (obj) {
-    return obj ? Object.keys(obj).map(function (key) {
-      var value = obj[key]
+  return []
+}
 
-      if (value === undefined) {
-        return ''
-      }
+/**
+ * Stringifies a query string, from an array of parameters
+ *
+ * @param parameters
+ * @returns {string}
+ */
+stringify = function (parameters) {
+  return parameters ? parameters.map(function (param) {
+    var key = decodeURIComponent(param.key)
+      , value = decodeURIComponent(param.value)
 
-      if (value === null) {
-        return encode(key)
-      }
+    if (value === undefined) {
+      return ''
+    }
 
-      if (Array.isArray(value)) {
-        var result = []
+    if (value === null) {
+      return encode(key)
+    }
 
-        value.slice().forEach(function (val2) {
-          if (val2 === undefined) {
-            return
-          }
+    if (Array.isArray(value)) {
+      var result = []
 
-          if (val2 === null) {
-            result.push(encode(key))
-          } else {
-            result.push(encode(key) + '=' + encode(val2))
-          }
-        })
+      value.slice().forEach(function (val2) {
+        if (val2 === undefined) {
+          return
+        }
 
-        return result.join('&')
-      }
+        if (val2 === null) {
+          result.push(encode(key))
+        } else {
+          result.push(encode(key) + '=' + encode(val2))
+        }
+      })
 
-      return encode(key) + '=' + encode(value)
-    }).filter(function (x) {
-      return x.length > 0
-    }).join('&') : ''
-  }
+      return result.join('&')
+    }
+
+    return encode(key) + '=' + encode(value)
+  }).filter(function (x) {
```

Comment:
seems like this does not retain empty key-value pairs.

- `a=&b`
- `a&b`
- `&a&b`
- `a&&b`
- `a&b&`